[max] implement Gemma3ForConditionalGeneration model

RossCampbellDev opened the pull request:
This PR adds the Gemma3ForConditionalGeneration model to the available MAX architectures, based on the technical report for multi modal Gemma3 and pre-existing implementations.
It is prepared for multi-gpu support, and tested against ChartQA

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/arch.py:
```diff
@@ -19,10 +19,10 @@
-    TextTokenizer,
+    TextAndVisionTokenizer,
 )
+from weight_adapters import convert_safetensor_state_dict
```
I think this needs to be `from .weight_adapters`

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/vision_model/projection.py:
```diff
@@ -0,0 +1,223 @@
+        self.tokens_per_side = int(config.mm_tokens_per_image**0.5)
+        self.kernel_size = self.patches_per_image // self.tokens_per_side
+
+    def __call__(self, vision_outputs: Tensor) -> TensorValue:
```
Both inputs and outputs should be TensorValue (they are all symbolic)

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/vision_model/projection.py:
```diff
@@ -0,0 +1,223 @@
+        normed_vision_outputs = self.mm_soft_emb_norm(pooled_vision_outputs)
+
+        projected_vision_outputs = matmul(
+            normed_vision_outputs, self.mm_input_projection_weight
```
`normed_vision_outputs @ self.mm_input_projection_weight`

RossCampbellDev replied:
for some reason i can't resolve this conversation - but the change has been made anyway

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/vision_model/projection.py:
```diff
@@ -0,0 +1,223 @@
+        then projects back to hidden size."""
+        x = self.fc1(x)
+        x = ops.gelu(
+            x, getattr(self.config.vision_config, "hidden_act", "none")
```
Do we need to use getattr here? the vision config used here is well typed (unlike HuggingFace's autoconfig object)

RossCampbellDev replied:
again, can't resolve but the change was made

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/model_config.py:
```diff
@@ -0,0 +1,306 @@
+    attention_dropout: float = 0.0
+    """The dropout ratio for the attention probabilities"""
+
+    model_type: str = "Gemma3"
```
I think this can be removed

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/model_config.py:
```diff
@@ -0,0 +1,306 @@
+    head_dim: int = 256
+    """The attention head dimension."""
+
+    model_type: str = "Gemma3ForConditionalGeneration"
```
This too can be removed

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/model.py:
```diff
@@ -98,66 +284,46 @@ def __init__(
+        )
 
-    @classmethod
+    @staticmethod
```
Making it a staticmethod makes sense, but for now we're stuck with classmethod because of the parent class (I think mypy would complain here)

k-w-w reviewed (commented):
Some drive by comments, great work!

RossCampbellDev committed:
Add MAX implementation of Gemma3ForConditionalGeneration

This commit adds the Gemma3ForConditionalGeneration model to MAX's
available pipeline architectures, with multi-GPU support enabled.

Language components have been adapted from the existing Gemma3 model,
while the 'Vision Tower' has been fully added as per the technical report.

The model has been benchmarked against ChartQA.

Co-authored-by: Ilya Savitsky ilya.savitsky@codethink.co.uk

tboerstad commented on max/python/max/pipelines/architectures/gemma3multimodal/model.py:
```diff
@@ -98,14 +284,22 @@ def __init__(
             return_logits,
         )
 
+        # # Initialize emtpy signal buffers
```
```suggestion
        # Initialize empty signal buffers
```

tboerstad commented on max/python/max/pipelines/architectures/gemma3multimodal/model.py:
```diff
@@ -14,36 +14,230 @@
 
 
-class Gemma3_MultiModalModel(Gemma3Model):
+# ⚠️ some parts are finished, some are possibly not (prepare vision inputs for example)
```
```suggestion
```

tboerstad commented on max/python/max/pipelines/architectures/gemma3multimodal/model.py:
```diff
@@ -171,22 +338,514 @@ def estimate_kv_cache_size(
+        final_images = _VisionStacker().stack(images)
+
+        tensor = Tensor.from_numpy(final_images)
+        tensor = _cast_to_dtype(
```
This is slow, since it creates an inference session under the hood. 
Can we do similar to InternVL? 

So in the tokenizer do:
`:rom max.pipelines.lib import float32_to_bfloat16_as_uint16`
 Store pixel_values as uint16 (bfloat16 bit representation)

 In model.py, reinterpret bytes:
  ```
tensor = Tensor.from_numpy(final_images)
tensor = tensor.view(DType.bfloat16, tensor.shape)
```

If that doesn't pan out, there's:
`from max.graph.tensor_utils import cast_dlpack_to` which does the same, but caches the inference session

tboerstad replied:
Sorry, on second thought let's just reuse `cast_dlpack_to`. 
This way there's minimal change to the code, `_cast_to_dtype` can be removed, and no risk of issues.

tboerstad commented on max/python/max/pipelines/architectures/gemma3multimodal/weight_adapters.py:
```diff
@@ -14,39 +14,63 @@
     state_dict: dict[str, Weights],
-    huggingface_config: AutoConfig,
-    **unused_kwargs,
+) -> tuple[dict[str, WeightData], dict[str, WeightData]]:
```
Can we change the return value to `-> dict[str, WeightData]`?
To be consistent with the other implementations

And rather use a prefix for the vision weights, ala:
```
GEMMA3_MULTIMODAL_SAFETENSOR_MAP: dict[str, str] = {
+    "vision_tower.vision_model.": "vision_tower.",
+    "multi_modal_": "vision_tower.", 
 }
```

RossCampbellDev replied:
might be a bit redundant - originally i had mimicked InternVL and had called the two conversion functions directly in `model.py#load_model()` to get my language+vision weights, which is fairly clean.

i then added the combined function in weight adapters that returns the tuple of dicts, which moves the function calls into weight adapters and allowed us to use `self.adapter()` in `model.py`

if i add a `vision_tower.` prefix to the vision weights, i need to then remove it again later otherwise I run into problems with missing weights and unexpected weights due to the names no longer matching.

unless there is some way to tell the various layers to expect a prefix string on their weights, it seems like the cleanest route is either calling the 2 conversion functions in `load_model()`, or doing what I currently have in this PR?

most others seem to use the former approach and ignore the `SupportedArchitecture`'s `adapter` entirely

k-w-w replied:
The weight adapter is only used internally by the `PipelineModel` (and we should honestly remove it from `SupportedArchitecture`), so the return type can be as flexible as needed. There might be issues with mypy with the current implementation, since the adapter must be: `WeightsAdapter = Callable[..., dict[str, WeightData]]`

I would recommend changing the return a single `dict`, and then splitting the weights when loading into the separate vision and language models

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/model_config.py:
```diff
@@ -0,0 +1,300 @@
+
+    hidden_act: str | None
+    """The non-linear activation function (function or string) in the encoder and pooler. If string, `"gelu"`,
+    `"relu"`, `"selu"` and `"gelu_new"` `"gelu"` are supported."""
```
"silu", also here's the list of supported activations: https://github.com/modular/modular/blob/d2dadf1a723d7e341f84d34e93223517b23bfca3/max/python/max/nn/linear.py#L1005-L1012

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/model_config.py:
```diff
@@ -0,0 +1,300 @@
+    `"relu"`, `"selu"` and `"gelu_new"` `"gelu"` are supported."""
+
+    hidden_size: int
+    """Dimensionality of the encoder layers and the pooler layer.  default 1152"""
```
Clean up the default values listed in the comments (since the defaults aren't actually set)

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/model_config.py:
```diff
@@ -0,0 +1,300 @@
+    """The size (resolution) of each image.  default 896"""
+
+    intermediate_size: int
+    """maybe not required? default 4304"""
```
Looks like this was used, can you update this comment?

k-w-w commented on max/python/max/pipelines/architectures/gemma3multimodal/model_config.py:
```diff
@@ -0,0 +1,300 @@
+    attention_bias: bool = False
+    """Whether to use a bias in the query, key, value and output projection layers during self-attention."""
+
+    attn_logit_softcapping: float = None
```
It looks like this isn't used (and I don't think we support softcapping), perhaps this should be removed

k-w-w reviewed (commented):
A few docstring comments, but the rest LGTM! Will let @tboerstad do the final approval

tboerstad commented on max/python/max/pipelines/architectures/gemma3multimodal/arch.py:
```diff
@@ -46,11 +45,8 @@
+    tokenizer=TextAndVisionTokenizer,
     default_weights_format=WeightsFormat.safetensors,
     multi_gpu_supported=True,
     rope_type=RopeType.normal,
```
```suggestion
    rope_type=RopeType.normal,
    required_arguments={
        "enable_prefix_caching": False,
        "enable_chunked_prefill": False,
    },
```

We do not support prefix_caching and chunked_prefill for VLMs at the moment. 
This fixes the issue with batch_size>1 failing

tboerstad commented on max/python/max/pipelines/architectures/gemma3multimodal/model.py:
```diff
@@ -171,22 +314,517 @@ def estimate_kv_cache_size(
+        Returns:
+            An initialized :obj:`KVCacheManager` instance.
+        """
+        return load_kv_manager(
```
```suggestion
        if available_cache_memory is not None:
            available_cache_memory = int(available_cache_memory * 0.9)

        return load_kv_manager(
```
Could you add this workaround? Otherwise our memory estimation fails, and we OOM

RossCampbellDev committed:
initial mypy cleanup

RossCampbellDev committed:
further mypy tweaks

the model functions but there are still mypy errors, and multi-device
functionality is incomplete

tboerstad commented on max/python/max/pipelines/architectures/gemma3multimodal/model.py:
```diff
@@ -171,22 +315,525 @@ def estimate_kv_cache_size(
+
+        final_images = _VisionStacker().stack(images)
+
+        tensor = Tensor.from_numpy(final_images)
```
```suggestion
```
This can be removed iiuc

RossCampbellDev committed:
last mypy fixes

RossCampbellDev committed:
restore legacy gemma3_multimodalmodel functionality

RossCampbellDev committed:
refactor VisionStacker for efficiency

RossCampbellDev committed:
move to flash attention
